### PR TITLE
Improve typedarray's map allocations

### PR DESCRIPTION
**Pull request by MaxGraey**
Not sure is it necessary optimization. @dcodeIO wdyt?

### CODE REVIEW

**Comment by dcodeIO** [comment]
Hmm, unlike `.filter` recently, the size of the array is known beforehand in `.map`, so the overhead here seems to be a single load of `.data`, right?

**Comment by MaxGraey** [comment]
Yes, but `instantiate<T>` internally call memory.fill which could skipped in this case. But yeah this expand size of generated bytecode

**Comment by DanielRX** [comment]
Does AS explicitly prevent .map being able to edit the original array (3rd param to the function)? Since in JS you have access to the array, is it possible to edit it (reducing the length maybe) which would be a problem?

**Comment by MaxGraey** [comment]
@DanielRX No, typed arrays hasn't "set" accessor for `length` so you can't change size of typed arrays after instantiation unlike `Array<T>`. At least legally (in safe style).

File: tests/compiler/std/typedarray.optimized.wat [hunk 1]
```diff
@@ -5512,36 +5512,44 @@
...
-  call $~lib/rt/pure/__retain
-  local.tee $5
+  local.tee $0
+  local.tee $2
+  if
+   local.get $2
+   call $~lib/rt/pure/__retain
+   drop
+   i32.const 0
+   call $~lib/rt/pure/__release
```
Comment (MaxGraey) [inline tests/compiler/std/typedarray.optimized.wat]: @dcodeIO Hmm, this shouldn't change code at all. But currently produce this extra retain/release block

**Comment by dcodeIO** [comment]
Good argument with `memory.fill`, and I agree that we should avoid `instantiate` altogether therefore, independently of shrink level. Shouldn't make a significant code size difference anyway since `instantiate` is basically inlined.

**Comment by MaxGraey** [comment]
I switched to manual allocs without `instantiate`. Conditional branch also produce side effect with unnecessary `retain/release`

**Comment by dcodeIO** [comment]
LGTM! Do you know if there are other uses of `instantiate<T>` in stdlib that can be optimized this way?

**Comment by MaxGraey** [comment]
It seems rest `instantiate<T>` have necessary usage

**Pull request merged**